ssl/quic/quic_channel.c

```diff
@@ -113,6 +114,88 @@ static int gen_rand_conn_id(OSSL_LIB_CTX *libctx, size_t len, QUIC_CONN_ID *cid)
     return 1;
 }
 
+static unsigned long chan_reset_token_hash(
+        const QUIC_STATELESS_RESET_TOKEN *tok)
+{
+    unsigned long h;
+
+    assert(sizeof(h) <= sizeof(tok->token));
+    memcpy(&h, tok->token, sizeof(h));
+    return h;
+}
+
+static int chan_reset_token_cmp(const QUIC_STATELESS_RESET_TOKEN *t1,
+                                const QUIC_STATELESS_RESET_TOKEN *t2)
+{
+    return CRYPTO_memcmp(t1->token, t2->token, sizeof(t1->token));
+}
+
+static int reset_token_obfuscate(QUIC_STATELESS_RESET_TOKEN *out,
+                                 const unsigned char *in, size_t len)
+{
+    /*
+     * TODO(QUIC): update this to AES encrypt the token in ECB mode with a
+     * random (per channel) key.
+     */
+    memcpy(out, in, sizeof(*out));
+    return 1;
+}
+
+/*
+ * Remove the old token from the table of legal tokens and insert the new
+ * one in it's place.
+ */
+static int chan_update_reset_token(QUIC_CHANNEL *ch,
+                                   const unsigned char *new, size_t new_len)
+{
+    QUIC_STATELESS_RESET_TOKEN *r;
+
+    /*
+     * Remove any existing reset token, if this fails it means there wasn't
+     * one and this can safely be ignored.
+     */
+    (void)lh_QUIC_STATELESS_RESET_TOKEN_delete(ch->stateless_reset_tokens,
+                                               &ch->cur_remote_reset_token);
+    reset_token_obfuscate(&ch->cur_remote_reset_token, new, new_len);
+    r = lh_QUIC_STATELESS_RESET_TOKEN_insert(ch->stateless_reset_tokens,
+                                             &ch->cur_remote_reset_token);
+    return r != NULL;
+}
+
+/*
+ * This is called by the demux whenever a new datagram arrives
+ */
+static int ch_stateless_reset_token_handler(const unsigned char *data,
+                                            size_t datalen, void *arg)
+{
+    QUIC_STATELESS_RESET_TOKEN tok;
+    QUIC_CHANNEL *ch = (QUIC_CHANNEL *)arg;
+
+    /*
+     * Perform some fast and cheap checks for a packet not being a stateless
+     * reset token.  RFC 9000 s. 10.3 specifies this layout for stateless
+     * reset packets:
+     *
+     *  Stateless Reset {
+     *      Fixed Bits (2) = 1,
+     *      Unpredictable Bits (38..),
+     *      Stateless Reset Token (128),
+     *  }
+     *
+     * We can rapidly check for the minimum length and that the first pair
+     * of bits in the first byte are 01.
+     *
+     * The funtion returns 1 if it is a stateless reset packet, 0 if it isn't
+     * and -1 if an error was encountered.
+     */
+    if (datalen < QUIC_STATELESS_RESET_TOKEN_LEN + 5 || (0300 & *data) != 0100)
```

Comment:
Was not expecting to see an octal Yoda condition. :laughing:

Reply:
For some bit masks, octal just makes more sense.  This is one.

Reply:
I've actually got the check wrong because we have to accept long format packets just in case.
What isn't clear is if we should accept any leading two bits in the packet.  The standard doesn't seem to be well defined.

> However, endpoints MUST treat any packet ending in a valid stateless reset token as a Stateless Reset, as other QUIC versions might allow the use of a long header.

This clearly implies that the `01` and `11` starts can be reset tokens.  It also kind of implies that any packet can be included undersized ones.  Elsewhere, there is mention of dropping short packets and there is no indication that leading bits of `00` and `10` can be reset packets.

Reply:
I would accept anything so long as bit 6 (the fixed bit) is set. i.e., I would allow either value for the Short Header bit.

This will be changed later at some point once we support RFC 9287, as it allows use of a different value for the 'fixed' bit to allow anti-ossification enforcement, etc. But we aren't implementing that yet, so it's fine to enforce the fixed bit for now.

Reply:
Cool, this is what I've done.